mono/mini/mini-exceptions.c

```diff
@@ -2953,6 +2953,43 @@ mono_print_thread_dump_internal (void *sigctx, MonoContext *start_ctx)
 	mono_runtime_stdout_fflush ();
 }
 
+typedef struct {
+	MonoInternalThread *thread;
+	GString *text;
+} ThreadDumpUserData;
+
+/* This needs to be async safe */
+static SuspendThreadResult
+get_thread_dump (MonoThreadInfo *info, gpointer ud)
+{
+	ThreadDumpUserData *user_data = (ThreadDumpUserData *)ud;
+	MonoInternalThread *thread = user_data->thread;
+
+	if (thread == mono_thread_internal_current ())
+		mono_get_eh_callbacks ()->mono_walk_stack_with_ctx (print_stack_frame_to_string, NULL, MONO_UNWIND_SIGNAL_SAFE, user_data->text);
+	else
+		mono_get_eh_callbacks ()->mono_walk_stack_with_state (print_stack_frame_to_string, mono_thread_info_get_suspend_state (info), MONO_UNWIND_SIGNAL_SAFE, user_data->text);
+
+	return MonoResumeThread;
+}
+
+const char*
+mono_print_target_thread_dump (void* threadPtr)
+{
+	ThreadDumpUserData ud;
+	MonoInternalThread* thread = (MonoInternalThread*)threadPtr;
+	ud.thread = thread;
+	ud.text = g_string_new_len (NULL, 20);
```

Comment:
Should we start with larger than 20? I am guessing this is usually a large string?

Reply:
Well this is going to be called on every sigsegv so I guess it may not always big a massive backtrace. Are you concerned about the perf hit from iteratively making the string larger? we could probably go bigger though, what do you think would be a good value?